File: sdk/python/mrpc.py
"""
mRPC Client — Python SDK

Usage:
    from mrpc import mRPC

    api = mRPC('http://localhost:8088/api.php')
    result = api.call('get_products', state='active')
    print(result['data'])
"""

import json
import time
import urllib.request
import urllib.error


class mRPC:
    """mRPC/1.0 Python Client"""

    def __init__(self, endpoint: str, token: str = None, timeout: int = 30):
        self.endpoint = endpoint
        self.token = token
        self.timeout = timeout
        self.version = 'mRPC/1.0'

    def call(self, cmd: str, **params) -> dict:
        """Вызвать команду.

        Args:
            cmd: Имя команды
            **params: Параметры команды

        Returns:
            dict с ответом сервера
        """
        body = {'cmd': cmd, **params}
        data = json.dumps(body).encode('utf-8')

        headers = {'Content-Type': 'application/json'}
        if self.token:
            headers['Authorization'] = f'Bearer {self.token}'

        req = urllib.request.Request(
            self.endpoint,
            data=data,
            headers=headers,
            method='POST'
        )

        start = time.time()
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as res:
                return json.loads(res.read().decode('utf-8'))
        except urllib.error.HTTPError as e:
            try:
                body = json.loads(e.read().decode('utf-8'))
                return body
            except Exception:
                return {
                    'ok': False,
                    'cmd': cmd,
                    'error': {'code': 'http', 'message': f'HTTP {e.code}: {e.reason}'},
                    'v': self.version,
                    'ms': round((time.time() - start) * 1000, 2)
                }
        except Exception as e:
            return {
                'ok': False,
                'cmd': cmd,
                'error': {'code': 'network', 'message': str(e)},
                'v': self.version,
                'ms': round((time.time() - start) * 1000, 2)
            }

    def help(self) -> dict:
        """Получить список команд."""
        return self.call('help')

    def describe(self, command: str) -> dict:
        """Описание конкретной команды."""
        return self.call('describe', command=command)

    def ping(self) -> dict:
        """Health check."""
        return self.call('ping')

    def login(self, email: str, password: str) -> dict:
        """Авторизация. Сохраняет токен если успешно."""
        res = self.call('auth_login', email=email, password=password)
        if res.get('ok') and res.get('data', {}).get('token'):
            self.token = res['data']['token']
        return res

    def set_token(self, token: str) -> 'mRPC':
        """Установить токен."""
        self.token = token
        return self


if __name__ == '__main__':
    import sys
    endpoint = sys.argv[1] if len(sys.argv) > 1 else 'http://localhost:8088/api.php'
    api = mRPC(endpoint)
    result = api.ping()
    print(json.dumps(result, indent=2, ensure_ascii=False))

File: sdk/js/mRPC.js
/**
 * mRPC Client — JavaScript SDK
 * 
 * @version 1.0
 * @example
 *   const api = new mRPC('http://localhost:8088/api.php');
 *   const products = await api.call('get_products', { state: 'active' });
 *   console.log(products.data);
 */
class mRPC {
  constructor(endpoint, options = {}) {
    this.endpoint = endpoint;
    this.token = options.token || null;
    this.timeout = options.timeout || 30000;
    this.version = 'mRPC/1.0';
    this.onError = options.onError || null;
  }

  /**
   * Вызвать команду
   * @param {string} cmd — имя команды
   * @param {object} params — параметры
   * @returns {Promise<object>} — ответ сервера
   */
  async call(cmd, params = {}) {
    const body = { cmd, ...params };
    const headers = { 'Content-Type': 'application/json' };
    if (this.token) headers['Authorization'] = `Bearer ${this.token}`;

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.timeout);

    try {
      const res = await fetch(this.endpoint, {
        method: 'POST',
        headers,
        body: JSON.stringify(body),
        signal: controller.signal,
      });

      const json = await res.json();

      if (!json.ok && this.onError) {
        this.onError(json.error, cmd, params);
      }

      return json;
    } catch (err) {
      const error = { ok: false, cmd, error: { code: 'network', message: err.message }, v: this.version };
      if (this.onError) this.onError(error.error, cmd, params);
      return error;
    } finally {
      clearTimeout(timer);
    }
  }

  /** Получить список команд */
  async help() {
    return this.call('help');
  }

  /** Описание конкретной команды */
  async describe(command) {
    return this.call('describe', { command });
  }

  /** Health check */
  async ping() {
    return this.call('ping');
  }

  /** Авторизация */
  async login(email, password) {
    const res = await this.call('auth_login', { email, password });
    if (res.ok && res.data?.token) {
      this.token = res.data.token;
    }
    return res;
  }

  /** Установить токен */
  setToken(token) {
    this.token = token;
    return this;
  }
}

// Node.js / CommonJS
if (typeof module !== 'undefined') module.exports = mRPC;

// ES Module
if (typeof window !== 'undefined') window.mRPC = mRPC;
